lib/nameMatcher.js

```diff
@@ -0,0 +1,307 @@
+'use strict';
+/**
+ * nameMatcher module
+ * exports nameMatcher object to expose methods for adding names and matching against them.
+ */
+
+
+// Private variables and external modules
+var rsvp = require('rsvp');
+// Set up rsvp global error handler to handle uncaught errors thrown within a promise-chain
+rsvp.on('error', function catchGlobalRSVPError(err) {
+    console.log('uncaught rejection of promise', err);
+    if (err instanceof Error && err.stack) {
+        console.error(err.stack);
+    }
+});
+var Promise = rsvp.Promise;
+var redis = require('redis');
+var natural = require('natural');
+var _ = require('underscore');
+
+var nameMatcher = {};
+var redisClient;
+// firstNames = e.g. {bob: bobMetaphoneHash, rachel: rachelMetaphoneHash}
+var firstNames;
+var lastNames;
+var firstNamesDev;
+var lastNamesDev;
+
+
+//static private helper functions
+/**
+ * Helper function to wrap standard callback-based async functions in a promise
+ * @param fn {function} The function to be called
+ * @param args {array} array of arguments to be passed to the function
+ * @param _this {object} (optional) object to use as `this` when calling the function defaults to null
+ * @return {Promise} promise to be resolved or rejected when the async fn is complete
+ */
+var promisify = function (fn, args, _this) {
+    _this = _this || null;
+    return new Promise(function promisifyWrapper(resolve, reject) {
+        var promiseCallback = function (err, result) {
+            if (err) {
+                reject(err);
+            } else {
+                resolve(result);
+            }
+            return;
+        };
+        args.push(promiseCallback);
+        return fn.apply(_this, args);
+    });
+};
+
+
+/**
+ * Helper function to generate the metaphone for a name.
+ * Will retrieve the metaphone from the local collection of names if it exists.
+ * @param name {string} the name to generate a metaphone for
+ * @param localCollection {object} a collection of name-indexed metaphones (optional)
+ * @return {array} an array of the generated metaphones.
+ */
+var getMetaphone = function (name, localCollection) {
+    localCollection = localCollection || {};
+    var dm = natural.DoubleMetaphone;
+    if (localCollection[name]) {
+        return localCollection[name];
+    }
+    return dm.process(name);
+};
+
+
+/**
+ * simple function to clear the database of names
+ */
+nameMatcher.clearNames = function() {
+    redisClient.del('firstNames', console.log('deleted first'));
```

Comment:
We are not going to use clearNames in prd, should be for dev-in-cli use only, I'll comment it as such